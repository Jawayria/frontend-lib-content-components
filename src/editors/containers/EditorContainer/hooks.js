import { useSelector } from 'react-redux';

import { RequestKeys } from '../../data/constants/requests';
import { selectors } from '../../data/redux';
import * as appHooks from '../../hooks';

export const {
  navigateCallback,
  nullMethod,
  saveBlock,
} = appHooks;

export const handleSaveClicked = ({ getContent, dispatch }) => (
  () => saveBlock({ content: getContent(), dispatch })
);
export const handleCancelClicked = ({ onClose }) => {
  if (onClose) {
    return onClose;
  }
  return navigateCallback(useSelector(selectors.app.returnUrl));
};
export const isInitialized = () => useSelector(selectors.app.isInitialized);
export const saveFailed = () => useSelector((state) => (
  selectors.requests.isFailed(state, { requestKey: RequestKeys.saveBlock })
));
