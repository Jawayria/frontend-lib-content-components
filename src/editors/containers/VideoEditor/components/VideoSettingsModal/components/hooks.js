import { useSelector } from 'react-redux';

import { keyStore } from '../../../../../utils';
import { actions, selectors } from '../../../../../data/redux';

export const selectorKeys = keyStore(selectors.video);

export const widgetValue = (key, dispatch) => ({
  formValue: useSelector(selectors.video[key]),
  setFormValue: (val) => dispatch(actions.video.load({ [key]: val })),
});

export default {
  selectorKeys,
  widgetValue,
};
