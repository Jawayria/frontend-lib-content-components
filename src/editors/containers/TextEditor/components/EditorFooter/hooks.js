import { useSelector } from 'react-redux';

import { RequestKeys } from '../../../../data/constants/requests';
import { selectors, thunkActions } from '../../../../data/redux';
import * as textEditorHooks from '../../hooks';

export const {
  navigateCallback,
  nullMethod,
  saveBlock,
} = textEditorHooks;

export const handleSaveClicked = ({ dispatch, editorRef }) => () => saveBlock({
  editorRef,
  saveBlockContent: (...args) => dispatch(thunkActions.app.saveBlock(...args)),
  returnUrl: useSelector(selectors.app.returnUrl),
});

export const handleCancelClicked = () => navigateCallback(useSelector(selectors.app.returnUrl));

export const isInitialized = () => useSelector(selectors.app.isInitialized);

export const saveFailed = () => useSelector((state) => (
  selectors.requests.isFailed(state, { requestKey: RequestKeys.saveBlock })
));

export const studioEndpointUrl = () => useSelector(selectors.app.studioEndpointUrl);
