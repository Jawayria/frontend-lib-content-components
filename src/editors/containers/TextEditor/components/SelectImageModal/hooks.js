import React from 'react';
import { useDispatch } from 'react-redux';

import { thunkActions } from '../../../../data/redux';
import * as module from './hooks';
import { sortFunctions, sortKeys } from './utils';

export const state = {
  images: (val) => React.useState(val),
  highlighted: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  return {
    searchString,
    onSearchChange: e => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: key => () => setSortBy(key),
  };
};

export const filteredList = ({ searchString, imageList }) => imageList.filter(
  ({ displayName }) => displayName.toLowerCase().includes(searchString.toLowerCase()),
);

export const displayList = ({ sortBy, searchString, images }) => module.filteredList({
  searchString,
  imageList: Object.values(images),
}).sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);

export const imgListHooks = ({
  setSelection,
  searchSortProps,
}) => {
  const dispatch = useDispatch();
  const [images, setImages] = module.state.images({});
  const [highlighted, setHighlighted] = module.state.highlighted(null);

  React.useEffect(() => {
    dispatch(thunkActions.app.fetchImages({ setImages }));
  }, []);

  return {
    images,
    // highlight by id
    selectBtnProps: {
      disabled: !highlighted,
      onClick: () => setSelection(images[highlighted]),
    },
    galleryProps: {
      highlighted,
      onHighlightChange: e => setHighlighted(e.target.value),
      displayList: module.displayList({ ...searchSortProps, images }),
    },
  };
};

export const fileInputHooks = ({ setSelection }) => {
  const dispatch = useDispatch();
  const ref = React.useRef();
  const click = () => ref.current.click();
  const addFile = (e) => {
    dispatch(thunkActions.app.uploadImage({
      file: e.target.files[0],
      setSelection,
    }));
  };

  return {
    click,
    addFile,
    ref,
  };
};

export const imgHooks = ({ setSelection }) => {
  const searchSortProps = module.searchAndSortHooks();
  const imgList = module.imgListHooks({ setSelection, searchSortProps });
  const fileInput = module.fileInputHooks({ setSelection });
  const { selectBtnProps, galleryProps } = imgList;

  return {
    fileInput,
    galleryProps,
    selectBtnProps,
    searchSortProps,
  };
};

export default {
  imgHooks,
};
