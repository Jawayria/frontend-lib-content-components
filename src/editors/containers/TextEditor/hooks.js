import { useState } from 'react';
import * as module from './hooks';

export const addImageUploadButton = (openModal, setImage) => (editor) => {
  editor.ui.registry.addButton('imageuploadbutton', {
    icon: 'image',
    onAction: openModal,
  });
  editor.ui.registry.addButton('editimagesettings', {
    icon: 'image',
    onAction() {
      const imgHTML = editor.selection.getNode();
      setImage({
        externalUrl: imgHTML.src,
        altText: imgHTML.alt,
        width: imgHTML.width,
        height: imgHTML.height,
      });
      openModal();
    },
  });
};

export const initializeEditorRef = (setRef, initializeEditor) => (editor) => {
  setRef(editor);
  initializeEditor();
};

// for toast onClose to avoid console warnings
export const nullMethod = () => {};

export const editorConfig = ({
  setEditorRef,
  blockValue,
  openModal,
  initializeEditor,
  setSelection,
}) => ({
  onInit: (evt, editor) => module.initializeEditorRef(setEditorRef, initializeEditor)(editor),
  initialValue: blockValue ? blockValue.data.data : '',
  init: {
    setup: module.addImageUploadButton(openModal, setSelection),
    plugins: 'link codesample emoticons table charmap code autoresize image imagetools',
    menubar: false,
    toolbar: 'undo redo | formatselect | '
      + 'bold italic backcolor | alignleft aligncenter '
      + 'alignright alignjustify | bullist numlist outdent indent |'
      + 'imageuploadbutton | link | emoticons | table | codesample | charmap |'
      + 'removeformat | hr |code',
    imagetools_toolbar: 'rotateleft rotateright | flipv fliph | editimagesettings',
    imagetools_cors_hosts: ['courses.edx.org'],
    height: '100%',
    content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:14px }',
    min_height: 1000,
    branding: false,
  },
});

export const modalToggle = () => {
  const [isOpen, setIsOpen] = useState(false);
  return {
    isOpen,
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  };
};
