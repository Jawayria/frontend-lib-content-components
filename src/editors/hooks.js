import { useEffect } from 'react';

import * as module from './hooks';

export const initializeApp = ({ initialize, data }) => useEffect(
  () => initialize(data),
  [],
);

export const navigateTo = (destination) => {
  window.location.assign(destination);
};

export const navigateCallback = (destination) => () => module.navigateTo(destination);
