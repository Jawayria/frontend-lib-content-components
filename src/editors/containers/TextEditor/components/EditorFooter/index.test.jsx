import React from 'react';
import { shallow } from 'enzyme';

import { formatMessage } from '../../../../../testUtils';
import * as hooks from './hooks';
import * as module from './index';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => jest.fn((...args) => ({ dispatch: args })).mockName('react-redux.dispatch'),
  useSelector: (selector) => ({ useSelector: selector }),
}));

jest.mock('./hooks', () => ({
  nullMethod: jest.fn().mockName('hooks.nullMethod'),
  handleCancelClicked: () => jest.fn().mockName('hooks.handleCancelClicked'),
  handleSaveClicked: (args) => ({ 'hooks.handleSaveClicked': args }),
  isInitialized: jest.fn(),
  saveFailed: jest.fn(),
}));

const { EditorFooter } = module;

describe('EditorFooter', () => {
  const props = {
    intl: { formatMessage },
    editorRef: jest.fn().mockName('args.editorRef'),
  };
  describe('snapshots', () => {
    test('renders as expected with default behavior', () => {
      hooks.isInitialized.mockReturnValueOnce(true);
      hooks.saveFailed.mockReturnValueOnce(false);
      expect(shallow(<EditorFooter {...props} />)).toMatchSnapshot();
    });
    test('not intialized, Spinner appears and button is disabled', () => {
      hooks.isInitialized.mockReturnValueOnce(false);
      hooks.saveFailed.mockReturnValueOnce(false);
      expect(shallow(<EditorFooter {...props} />)).toMatchSnapshot();
    });
    test('Save Failed, error message raised', () => {
      hooks.isInitialized.mockReturnValueOnce(true);
      hooks.saveFailed.mockReturnValueOnce(true);
      expect(shallow(<EditorFooter {...props} />)).toMatchSnapshot();
    });
  });
});
