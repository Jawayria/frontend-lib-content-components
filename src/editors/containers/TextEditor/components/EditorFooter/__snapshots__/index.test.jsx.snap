// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorFooter snapshots Save Failed, error message raised 1`] = `
<div
  className="editor-footer mt-auto"
>
  <Toast
    onClose={[MockFunction hooks.nullMethod]}
    show={true}
  >
    <FormattedMessage
      defaultMessage="Error: Content save failed. Try again later."
      description="Error message displayed when content fails to save."
      id="authoring.editorfooter.save.error"
    />
  </Toast>
  <ModalDialog.Footer>
    <ActionRow>
      <ActionRow.Spacer />
      <Button
        aria-label="Discard Changes and Return to Learning Context"
        onClick={[MockFunction hooks.handleCancelClicked]}
        variant="tertiary"
      >
        <FormattedMessage
          defaultMessage="Cancel"
          description="Label for cancel button"
          id="authoring.editorfooter.cancelButton.label"
        />
      </Button>
      <Button
        aria-label="Save Changes and Return to Learning Context"
        disabled={false}
        onClick={
          Object {
            "hooks.handleSaveClicked": Object {
              "dispatch": [MockFunction react-redux.dispatch],
              "editorRef": [MockFunction args.editorRef],
            },
          }
        }
      >
        <FormattedMessage
          defaultMessage="Save"
          description="Label for Save button"
          id="authoring.editorfooter.savebutton.label"
        />
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</div>
`;

exports[`EditorFooter snapshots not intialized, Spinner appears and button is disabled 1`] = `
<div
  className="editor-footer mt-auto"
>
  <ModalDialog.Footer>
    <ActionRow>
      <ActionRow.Spacer />
      <Button
        aria-label="Discard Changes and Return to Learning Context"
        onClick={[MockFunction hooks.handleCancelClicked]}
        variant="tertiary"
      >
        <FormattedMessage
          defaultMessage="Cancel"
          description="Label for cancel button"
          id="authoring.editorfooter.cancelButton.label"
        />
      </Button>
      <Button
        aria-label="Save Changes and Return to Learning Context"
        disabled={true}
        onClick={
          Object {
            "hooks.handleSaveClicked": Object {
              "dispatch": [MockFunction react-redux.dispatch],
              "editorRef": [MockFunction args.editorRef],
            },
          }
        }
      >
        <Spinner
          animation="border"
          className="mr-3"
        />
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</div>
`;

exports[`EditorFooter snapshots renders as expected with default behavior 1`] = `
<div
  className="editor-footer mt-auto"
>
  <ModalDialog.Footer>
    <ActionRow>
      <ActionRow.Spacer />
      <Button
        aria-label="Discard Changes and Return to Learning Context"
        onClick={[MockFunction hooks.handleCancelClicked]}
        variant="tertiary"
      >
        <FormattedMessage
          defaultMessage="Cancel"
          description="Label for cancel button"
          id="authoring.editorfooter.cancelButton.label"
        />
      </Button>
      <Button
        aria-label="Save Changes and Return to Learning Context"
        disabled={false}
        onClick={
          Object {
            "hooks.handleSaveClicked": Object {
              "dispatch": [MockFunction react-redux.dispatch],
              "editorRef": [MockFunction args.editorRef],
            },
          }
        }
      >
        <FormattedMessage
          defaultMessage="Save"
          description="Label for Save button"
          id="authoring.editorfooter.savebutton.label"
        />
      </Button>
    </ActionRow>
  </ModalDialog.Footer>
</div>
`;
